usersession/client/client.go

```diff
@@ -0,0 +1,244 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2015-2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"strconv"
+	"sync"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+)
+
+// dialSessionAgent connects to a user's session agent
+//
+// The host portion of the address is interpreted as the numeric user
+// ID of the target user.
+func dialSessionAgent(network, address string) (net.Conn, error) {
+	host, _, err := net.SplitHostPort(address)
+	if err != nil {
+		return nil, err
+	}
+	socket := filepath.Join(dirs.XdgRuntimeDirBase, host, "snapd-session-agent.socket")
+	return net.Dial("unix", socket)
+}
+
+type Client struct {
+	doer *http.Client
+}
+
+func New() *Client {
+	transport := &http.Transport{Dial: dialSessionAgent, DisableKeepAlives: true}
+	return &Client{
+		doer: &http.Client{Transport: transport},
+	}
+}
+
+type Error struct {
+	Kind    string      `json:"kind"`
+	Value   interface{} `json:"value"`
+	Message string      `json:"message"`
+}
+
+func (e *Error) Error() string {
+	return e.Message
+}
+
+type response struct {
+	// Not from JSON
+	uid        int
+	err        error
+	statusCode int
+
+	Result json.RawMessage `json:"result"`
+	Type   string          `json:"type"`
+}
+
+func (resp *response) checkError() {
+	if resp.Type != "error" {
+		return
+	}
+	var resultErr Error
+	err := json.Unmarshal(resp.Result, &resultErr)
+	if err != nil || resultErr.Message == "" {
+		resp.err = fmt.Errorf("server error: %q", http.StatusText(resp.statusCode))
+	} else {
+		resp.err = &resultErr
+	}
+}
+
+func (client *Client) do(ctx context.Context, method, urlpath string, query url.Values, headers map[string]string, body []byte) ([]*response, error) {
```

Comment:
Nitpick, I would document this method and give it a slightly different name to differentiate it from `client.Client.go`, perhaps `doMany`? I will leave it up to you to decide.

Reply:
yes, something like doMany seems appropriate given the []*response result

Reply:
It is doMany in the current version of the branch.  I forgot to mark this resolved.